cpp/dolfinx/fem/CoordinateElement.cpp

```diff
@@ -86,8 +88,106 @@ void CoordinateElement::compute_reference_geometry(
+
+    // Compute Jacobian and inverse
+    _evaluate_basis_derivatives(dphi.data(), 1, 1, X0.data());
+    Eigen::Matrix<double, Eigen::Dynamic, Eigen::Dynamic, Eigen::RowMajor, 3, 3>
+        J0(_gdim, _tdim);
+    J0 = cell_geometry.matrix().transpose() * dphi;
+    Eigen::Matrix<double, Eigen::Dynamic, Eigen::Dynamic, Eigen::RowMajor, 3, 3>
+        K0(_tdim, _gdim);
+    if (_gdim == _tdim)
+    {
+      K0 = J0.inverse();
+      detJ.fill(J0.determinant());
+    }
+    else
+    {
+      // Penrose-Moore pseudo-inverse
+      K0 = (J0.transpose() * J0).inverse() * J0.transpose();
+      detJ.fill(std::sqrt((J0.transpose() * J0).determinant()));
+    }
+
+    // Fill result for J, K and detJ
+    Eigen::Map<
+        Eigen::Matrix<double, Eigen::Dynamic, Eigen::Dynamic, Eigen::RowMajor>>
+        Jview(J.data(), _gdim * num_points, _tdim);
+    Jview = J0.replicate(num_points, 1);
+    Eigen::Map<
+        Eigen::Matrix<double, Eigen::Dynamic, Eigen::Dynamic, Eigen::RowMajor>>
+        Kview(K.data(), _tdim * num_points, _gdim);
+    Kview = K0.replicate(num_points, 1);
+
+    // Calculate X for each point
+    for (int ip = 0; ip < num_points; ++ip)
+      X.row(ip) = K0 * (x.row(ip).matrix().transpose() - x0);
+  }
+  else
+  {
+    // Newton's method for non-affine geometry
+    Eigen::Matrix<double, Eigen::Dynamic, 1, Eigen::ColMajor, 3, 1> xk(x.cols(),
+                                                                       1);
+    Eigen::Matrix<double, Eigen::Dynamic, 1, Eigen::ColMajor, 3, 1> Xk(_tdim,
+                                                                       1);
+
+    for (int ip = 0; ip < num_points; ++ip)
+    {
+      Eigen::Map<Eigen::Matrix<double, Eigen::Dynamic, Eigen::Dynamic,
+                               Eigen::RowMajor>>
+          Jview(J.data() + ip * _gdim * _tdim, _gdim, _tdim);
+      Eigen::Map<Eigen::Matrix<double, Eigen::Dynamic, Eigen::Dynamic,
+                               Eigen::RowMajor>>
+          Kview(K.data() + ip * _gdim * _tdim, _tdim, _gdim);
+      // TODO: Xk - use cell midpoint instead?
+      Xk.setZero();
+      const int max_its = 10;
+      int k;
+      for (k = 0; k < max_its; ++k)
+      {
+        // Compute physical coordinates
+        _evaluate_basis_derivatives(phi.data(), 0, 1, Xk.data());
+        xk = cell_geometry.matrix().transpose() * phi;
+
+        // Compute Jacobian and inverse
+        _evaluate_basis_derivatives(dphi.data(), 1, 1, Xk.data());
+        Jview = cell_geometry.matrix().transpose() * dphi;
+        if (_gdim == _tdim)
+          Kview = Jview.inverse();
+        else
+          // Penrose-Moore pseudo-inverse
+          Kview = (Jview.transpose() * Jview).inverse() * Jview.transpose();
+
+        // Increment to new point in reference
+        Eigen::Matrix<double, Eigen::Dynamic, 1, Eigen::ColMajor, 3, 1> dX
+            = Kview * (x.row(ip).matrix().transpose() - xk);
+        if (dX.squaredNorm() < 1e-12)
```

Comment:
Similar comment here. As we have chosen a given tolerance, should we give the user the ability to control it?